### PR TITLE
oop-special_class_method-medium-q8

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/special_class_method/medium/q8/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        Object[][] arguments = {
+                {skill, String.class},
+                {bounty, int.class}
+        };
+        Object testInstance = testClass.createInstance(arguments);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName1, testInstance);
+        assertTrue(testClass.hasModifier(getAttributeMethodName1, null, "public"),
+                "Your " + getAttributeMethodName1 + " method does not have the correct visibility modifier.");
+        _assertEquals(skill, getMethodOutput,
+                "Your " + getAttributeMethodName1 + " method does not return the value of the size field.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/MainTest.java:103]:
```suggestion
                "Your " + getAttributeMethodName1 + " method does not return the value of the " + attributeName1 + " attribute.");
```

Comment (SamamthaH) [reply src/oop/special_class_method/medium/q8/MainTest.java:103]: also change any other instances of field to attribute

File: src/oop/special_class_method/medium/q8/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# oop-special_class_method-medium-q7
+
+## Question Text
+
+You want to make a battle game with characters that are modeled after the characters from the well-known anime called 
+One Piece who ate the devil fruit to gain a special ability. Make a Java class called **Luffy** that has several
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/QuestionText.md:6]:
```suggestion
One Piece who ate the devil fruit to gain a special ability. Make a Java class called **Luffy** that has  
```

File: src/oop/special_class_method/medium/q8/QuestionText.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+Your `Luffy` class must have the following items
+
+- Attributes:
+    - An `String` called **skill**
+    - An `int` called **bounty**
+
+- Constructor:
+    - a `Luffy` constructor that is visible to other classes and accepts a `String` and an `int` and uses it to initialize the
+      **skill** and **bounty** fields.
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/QuestionText.md:22]: change the "field" to "attribute" in this file as well

File: src/oop/special_class_method/medium/q8/QuestionText.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+You want to make a battle game with characters that are modeled after the characters from the well-known anime called 
+One Piece who ate the devil fruit to gain a special ability. Make a Java class called **Luffy** that has several 
+attributes called  **skill** and **bounty**. This class has the basic constructor that initializes the attributes, 
+accessors and mutators for each attribute, and a **toString()** method that returns the total amount of money available 
+as a String.
+
+
+### Luffy Class:
+
+Your `Luffy` class must have the following items
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/QuestionText.md:14]: change this to reflect that the following things will have been provided

File: src/oop/special_class_method/medium/q8/QuestionText.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      **skill** and **bounty** fields.
+
+You need to add the following items:
+
+- Methods:
+    - a `setSkill` method which accepts a `String` and uses it to set the value of the **skill** field
+    - a `getSkill` method which returns the value stored in the **skill** field
+    - a `setBounty` method which accepts an `int` and uses it to set the value of the **bounty** field
+    - a `getBounty` method which returns the value stored in the **bounty** field
+    - a `toString` method which returns the string `"My special skill is " + skill + " and my bounty is " + bounty`.
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/QuestionText.md:31]: use x y format

File: src/oop/special_class_method/medium/q8/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        };
+        Object testInstance = testClass.createInstance(arguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, int.class}
+        };
+        testClass.callMethod(setAttributeMethodName2, setMethodArguments, testInstance);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName2, testInstance);
+        _assertEquals(updatedValue, getMethodOutput, errorMessage);
+    }
+}
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/MainTest.java:217]: EOF line

File: src/oop/special_class_method/medium/q8/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName1, testInstance);
+        assertTrue(testClass.hasModifier(getAttributeMethodName1, null, "public"),
+                "Your " + getAttributeMethodName1 + " method does not have the correct visibility modifier.");
+        _assertEquals(skill, getMethodOutput,
+                "Your " + getAttributeMethodName1 + " method does not return the value of the size field.");
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void correctGetNickelMethod(String skill, int bounty) throws Throwable {
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/MainTest.java:108]:
```suggestion
    public void correctGetBountyMethod(String skill, int bounty) throws Throwable {
```

File: src/oop/special_class_method/medium/q8/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        Object[][] arguments = {
+                {skill, String.class},
+                {bounty, int.class}
+        };
+        Object testInstance = testClass.createInstance(arguments);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName2, testInstance);
+        assertTrue(testClass.hasModifier(getAttributeMethodName2, null, "public"),
+                "Your " + getAttributeMethodName2 + " method does not have the correct visibility modifier.");
+        _assertEquals(bounty, getMethodOutput,
+                "Your " + getAttributeMethodName2 + " method does not return the value of the size field.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/MainTest.java:118]:
```suggestion
                "Your " + getAttributeMethodName2 + " method does not return the value of the " + attributeName2 + " attribute.");
```

File: src/oop/special_class_method/medium/q8/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+                {bounty, int.class}
+        };
+        Object testInstance = testClass.createInstance(instantiationArguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, String.class}
+        };
+        Class<?>[] methodModifierClasses = {
+                String.class
+        };
+        testClass.callMethod(setAttributeMethodName1, setMethodArguments, testInstance);
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/MainTest.java:135]: this should probably be two tests, one for name and one for arguments

File: src/oop/special_class_method/medium/q8/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+                {initialValue, int.class}
+        };
+        Object testInstance = testClass.createInstance(instantiationArguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, int.class}
+        };
+        Class<?>[] methodModifierClasses = {
+                int.class
+        };
+        testClass.callMethod(setAttributeMethodName2, setMethodArguments, testInstance);
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/MainTest.java:158]: this should probably be two tests, one for name and one for arguments

File: src/oop/special_class_method/medium/q8/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+                {value2, int.class}
+        };
+        String incorrectToStringMessage = "Your toString method does not return the correct String.";
+        Object testInstance = testClass.createInstance(arguments);
+        String expectedOutput = "My special skill is " + value1 + " and my bounty is " + value2;
+        Object toStringOutput = testClass.callMethod("toString", testInstance);
+        _assertEquals(expectedOutput, toStringOutput, incorrectToStringMessage);
+    }
+
+    @ParameterizedTest
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/MainTest.java:182]: I'd say combine this and the second one into one methods work together tests that also includes the toString

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/medium/q8/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+        _assertEquals("initial string", getMethodOutput,
+                "Your " + getAttributeMethodName1 + " method does not return the value of the " + attributeName1 + " attribute.");
+        getMethodOutput = testClass.callMethod(getAttributeMethodName2, testInstance);
+        _assertEquals(123, getMethodOutput,
+                "Your " + getAttributeMethodName2 + " method does not return the value of the " + attributeName2 + " attribute.");
+        Object[][] setMethodArguments = {
+                {value1, String.class}
+        };
+        Object setMethodOutput = testClass.callMethod(setAttributeMethodName1, setMethodArguments, testInstance);
+        _assertEquals(value1, testClass.getFieldValue(testInstance, attributeName1),
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/MainTest.java:244]: this should instead be using the get method

File: src/oop/special_class_method/medium/q8/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+                "Your " + setAttributeMethodName1 + " method does not correctly update the value of " + attributeName1 + " after using the " + getAttributeMethodName1 + " method.");
+        assertNull(setMethodOutput, String.join(" ", "Your", setAttributeMethodName1, "method should not return any output"));
```
Comment (SamamthaH) [inline src/oop/special_class_method/medium/q8/MainTest.java:246]: testing for no returned output should not happen here

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by SamamthaH — APPROVED** [review]

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
See feedback on other questions in this category

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**